app/scripts/lib/rpc-method-middleware/handlers/wallet-requestPermissions.ts

```diff
@@ -0,0 +1,265 @@
+    (Object.keys(requestedPermissions).length === 0 &&
+      !haveLegacyPermissions) ||
+    Object.keys(requestedPermissions).length > 0
+  ) {
+    const [_grantedPermissions] = await requestPermissionsForOrigin(
+      requestedPermissions,
+    );
+    // permissions are frozen and must be cloned before modified
+    grantedPermissions = { ..._grantedPermissions };
+  }
+
+  if (legacyApproval) {
+    let newCaveatValue = {
+      requiredScopes: {},
+      optionalScopes: {},
+      isMultichainOrigin: false,
+    };
+
+    if (isSnapId(origin)) {
+      newCaveatValue.optionalScopes = {
+        'wallet:eip155': {
+          accounts: [],
+        },
+      };
+    } else {
+      newCaveatValue = setPermittedEthChainIds(
+        newCaveatValue,
+        legacyApproval.approvedChainIds,
+      );
+    }
+
+    newCaveatValue = setEthAccounts(
+      newCaveatValue,
+      legacyApproval.approvedAccounts,
+    );
+
+    const permissions = getPermissionsForOrigin() || {};
+    let caip25Endowment = permissions[Caip25EndowmentPermissionName];
+    const existingCaveatValue = caip25Endowment?.caveats?.find(
+      ({ type }) => type === Caip25CaveatType,
+    )?.value as Caip25CaveatValue;
+    if (existingCaveatValue) {
+      if (existingCaveatValue.isMultichainOrigin) {
+        return end(
+          new Error(
+            'Cannot modify permission granted via the Multichain API. Either modify the permission using the Multichain API or revoke permissions and request again.',
+          ),
+        );
+      }
+
+      updateCaveat(
+        origin,
+        Caip25EndowmentPermissionName,
+        Caip25CaveatType,
+        newCaveatValue,
+      );
+    } else {
+      caip25Endowment = grantPermissions({
+        subject: { origin },
+        approvedPermissions: {
+          [Caip25EndowmentPermissionName]: {
+            caveats: [
+              {
+                type: Caip25CaveatType,
+                value: newCaveatValue,
+              },
+            ],
+          },
+        },
+      })[Caip25EndowmentPermissionName];
+    }
+
+    // We cannot derive ethAccounts directly from the CAIP-25 permission
+    // because the accounts will not be in order of lastSelected
+    const ethAccounts = getAccounts();
+
+    grantedPermissions[RestrictedMethods.eth_accounts] = {
+      ...caip25Endowment,
+      parentCapability: RestrictedMethods.eth_accounts,
+      caveats: [
+        {
+          type: CaveatTypes.restrictReturnedAccounts,
+          value: ethAccounts,
+        },
+      ],
+    };
+
+    if (!isSnapId(origin)) {
+      grantedPermissions[PermissionNames.permittedChains] = {
+        ...caip25Endowment,
+        parentCapability: PermissionNames.permittedChains,
+        caveats: [
+          {
+            type: CaveatTypes.restrictNetworkSwitching,
+            value: legacyApproval.approvedChainIds,
+          },
+        ],
+      };
+    }
+  }
+
+  res.result = Object.values(grantedPermissions) as Json;
```

Comment:
These types don't quite match; it seems that `Object.values(grantedPermissions)` can contain `undefined` values. Although `grantedPermissions` was declared up front as `GrantedPermissions`, it seems that TypeScript has narrowed the type to `Partial<...>` because of what `requestPermissionsFromOrigin` returns.

If it's important to maintain JSON compatibility, perhaps we can remove `undefined` from this object? Then we can type-cast it to `GrantedPermissions`.

``` suggestion
  res.result = Object.values(grantedPermissions).filter(
    (value) => value !== undefined,
  ) as Json;
```